pkg/segment/utils/number.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2021-2024 SigScalr, Inc.
+//
+// This file is part of SigLens Observability Solution
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/siglens/siglens/pkg/utils"
+)
+
+type numberType = byte
+
+const (
+	invalidType numberType = iota
+	int64Type
+	float64Type
+	backfillType
+)
+
+type Number struct {
+	bytes [9]byte // 8 bytes for the number; last byte for the type
+}
+
+func (n *Number) SetInvalidType() {
+	n.bytes[8] = invalidType
+}
+
+func (n *Number) SetBackfillType() {
+	n.bytes[8] = backfillType
+}
+
+func (n *Number) SetInt64(i int64) {
+	utils.Int64ToBytesLittleEndianInplace(i, n.bytes[:8])
+	n.bytes[8] = int64Type
+}
+
+func (n *Number) SetFloat64(f float64) {
+	utils.Float64ToBytesLittleEndianInplace(f, n.bytes[:8])
+	n.bytes[8] = float64Type
+}
+
+func (n *Number) Int64() (int64, error) {
+
+	if n.bytes[8] == backfillType {
+		return 0, nil
+	}
+
+	if n.bytes[8] != int64Type {
+		return 0, fmt.Errorf("Not a int64, t: %v", n.bytes[8])
+	}
+
+	return utils.BytesToInt64LittleEndian(n.bytes[:8]), nil
+}
+
+func (n *Number) Float64() (float64, error) {
+
+	if n.bytes[8] == backfillType {
+		return 0, nil
+	}
+
+	if n.bytes[8] != float64Type {
+		return 0, fmt.Errorf("Not a float64, t: %v", n.bytes[8])
+	}
+
+	return utils.BytesToFloat64LittleEndian(n.bytes[:8]), nil
+}
+
+func ConvertBytes(buf []byte) (int64, float64, SS_DTYPE) {
```

Comment:
Should this be a method on `Number`?

Reply:
nit: rename `ConvertBytes`. I don't really know what that means without reading the whole function

Reply:
will switch to `ConvertBytesToNumber` , the caller of this function needs to get back the int64, float64 and the type, since the rest of the code where it uses these return variables need that. And there is lot of changes if we switch this to be a method on `Number`.
eventually we would get rid of the `Number` class and use the `NumTypeEnclosure` which was actually used for this purpose.